packages/middleware-retry/src/defaultStrategy.spec.ts

```diff
@@ -133,71 +354,34 @@ describe("defaultStrategy", () => {
-
-      const mockError = new Error("mockError");
-      const next = jest.fn().mockRejectedValueOnce(mockError);
-
-      const retryStrategy = new StandardRetryStrategy(maxAttempts);
-      try {
-        await retryStrategy.retry(next, {} as any);
-      } catch (error) {
-        expect(error).toStrictEqual(mockError);
-      }
-
+      const mockError = new Error();
+      await mockFailedOperation(maxAttempts, { mockError });
       expect(defaultRetryDecider as jest.Mock).toHaveBeenCalledTimes(1);
       expect(defaultRetryDecider as jest.Mock).toHaveBeenCalledWith(mockError);
     });
 
-    it("when the the maximum number of attempts is reached", async () => {
-      const mockError = new Error("mockError");
-      const next = jest.fn().mockRejectedValue(mockError);
-
-      const retryStrategy = new StandardRetryStrategy(maxAttempts);
-      try {
-        await retryStrategy.retry(next, {} as any);
-      } catch (error) {
-        expect(error).toStrictEqual(mockError);
-      }
+    it("when the maximum number of attempts is reached", async () => {
+      await mockFailedOperation(maxAttempts);
       expect(defaultRetryDecider as jest.Mock).toHaveBeenCalledTimes(
         maxAttempts - 1
       );
     });
-  });
-
-  it("should delay equal to the value returned by delayDecider", async () => {
-    jest.spyOn(global, "setTimeout");
-
-    const FIRST_DELAY = 100;
-    const SECOND_DELAY = 200;
 
-    (defaultDelayDecider as jest.Mock)
-      .mockReturnValueOnce(FIRST_DELAY)
-      .mockReturnValueOnce(SECOND_DELAY);
-
-    const mockError = new Error("mockError");
-    const next = jest.fn().mockRejectedValue(mockError);
-
-    const retryStrategy = new StandardRetryStrategy(3);
-    try {
-      await retryStrategy.retry(next, {} as any);
-    } catch (error) {
-      expect(error).toStrictEqual(mockError);
-      expect(error.$metadata.totalRetryDelay).toEqual(
-        FIRST_DELAY + SECOND_DELAY
-      );
-    }
-
-    expect(defaultDelayDecider as jest.Mock).toHaveBeenCalledTimes(2);
-    expect(setTimeout).toHaveBeenCalledTimes(2);
-    expect(setTimeout).toHaveBeenNthCalledWith(
-      1,
-      expect.any(Function),
-      FIRST_DELAY
-    );
-    expect(setTimeout).toHaveBeenNthCalledWith(
-      2,
-      expect.any(Function),
-      SECOND_DELAY
-    );
+    it("when retryQuota.hasRetryTokens returns false", async () => {
+      const {
+        hasRetryTokens,
+        retrieveRetryTokens,
+        releaseRetryTokens
+      } = getDefaultRetryQuota();
+      (hasRetryTokens as jest.Mock).mockReturnValueOnce(false);
+
+      const mockError = new Error();
+      await mockFailedOperation(maxAttempts, { mockError });
+
+      expect(hasRetryTokens).toHaveBeenCalledTimes(1);
+      expect(hasRetryTokens).toHaveBeenCalledWith(mockError);
+      expect(retrieveRetryTokens).not.toHaveBeenCalled();
+      expect(releaseRetryTokens).not.toHaveBeenCalled();
+    });
```

Comment:
The tests looks good to me. Can you add the 2 test cases for retry quota in the design doc?

Reply:
Will do!

Reply:
Will do

Reply:
Added in [commit](https://github.com/aws/aws-sdk-js-v3/pull/1258/commits/8a1c222d6a81e216cae6871f706698aa68620b3c)